docs/generate/dataplane-customization.md

```diff
@@ -0,0 +1,96 @@
+
+## Change Client Behavior
+
+### Support Azure Active Directory Authentication
+
+See [Authentication in SDK](https://github.com/Azure/autorest/blob/main/docs/generate/authentication.md).
+
+It is advised to update OpenAPI specs with `security: AADToken`. And then re-generate the SDK.
+
+If one does not wish to change the specs, add following configure to README_SPEC.md, and re-generate the SDK.
+```
+security: AADToken
+security-scopes: https://cognitiveservices.azure.com/.default
+```
+Here is an [example](https://github.com/weidongxu-microsoft/azure-sdk-for-java/commit/d30773a776f36e8269c7b2377e1d948fad2f5b82) of the modification of README_SPEC.md and the resulted code changes.
+Note that a new method `credential(TokenCredential tokenCredential)` is generated in builder class.
+
+### Support Customized Authentication
+
+Usually `security: AzureKey` can provide implementation of API key authentication.
+
+For more complex API key authentication, one can customize the builder class.
+
+Here is an [example](https://github.com/weidongxu-microsoft/azure-sdk-for-java/commit/f5337b6fbc5937683e01299a509b505e01d8b7ce) of the modification of builder to support subscription key and API key.
+1. Create `MetricsAdvisorKeys` and `MetricsAdvisorKeyCredential` class in `models` package.
+2. Modify `MetricsAdvisorClientBuilder` to accept `MetricsAdvisorKeyCredential` from user, and then pass the keys in it to HTTP pipeline.
+
+Note that the `Generated` annotation on `createHttpPipeline()` is removed, as we customized this method.
+This ensures that future code generate will not override this customized method.
+
+## Change Operation Behavior
+
+Generated client methods accept `ReqestOptions` for request parameters.
+For frequently invoked methods, one can customize them for better user-experience.
+
+Here is an [example](https://github.com/weidongxu-microsoft/azure-sdk-for-java/commit/c3e8ebc1bf13dd29d67da0b68d23a84197ce681e) of the modification of client to support options for client methods.
+
+1. Create `ListDataFeedFilter` and `ListDataFeedOptions` class in `models` package.
+2. Modify `MetricsAdvisorAsyncClient` and `MetricsAdvisorClient` to provide overloaded `listDataFeeds` method that takes `ListDataFeedOptions` options.
+3. Convert options from `ListDataFeedOptions` to `RequestOptions` in `listDataFeeds` method.
+
+Note that the conversion only happen once in package private `MetricsAdvisorAsyncClient.listDataFeeds(ListDataFeedOptions, Context)` method, which is invoked by all added `listDataFeeds` methods.
+
+## Add Model
+
+Generated client methods accept and produce `BinaryData` for request payload and response body.
+For frequently invoked methods, one can customize them for better user-experience.
+
+AutoRest Java `generate-models` option is provided to generate models from OpenAPI specs, to `implmenetation.models` package. They are provided only as reference.
```

Comment:
Will it generate all models?  What if I only need a subset of all models?

Reply:
It will. But this is only as reference. If you check metricsadvisor, the `DataFeed` model written by hand is completely different to the model from swagger.

So if service want to use some of the generated models, they can use IDE to move them to the public `models` package. But the advise is to review them first before doing so.

Reply:
In this section and example, I only want to demonstrate how to convert from/to model, not the design of user-facing class. Hence I only use the swagger generated `DataFeedDetail` class.